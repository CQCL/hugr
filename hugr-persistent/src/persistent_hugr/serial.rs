//! Serialized format for [`PersistentHugr`]

use hugr_core::Hugr;

use crate::{CommitStateSpace, SerdeHashResolver, state_space::serial::SerialCommitStateSpace};

use super::PersistentHugr;

/// Serialized format for [`PersistentHugr`]
#[derive(Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
pub struct SerialPersistentHugr<H> {
    /// The state space of all commits.
    state_space: SerialCommitStateSpace<H>,
}

impl<H: Into<Hugr> + From<Hugr> + serde::Serialize> PersistentHugr<SerdeHashResolver<H>> {
    /// Create a new [`CommitStateSpace`] from its serialized format
    pub fn from_serial(value: SerialPersistentHugr<H>) -> Self {
        let SerialPersistentHugr { state_space } = value;
        let state_space = CommitStateSpace::from_serial(state_space);
        Self { state_space }
    }

    /// Convert a [`CommitStateSpace`] into its serialized format
    pub fn into_serial(self) -> SerialPersistentHugr<H> {
        let Self { state_space } = self;
        let state_space = state_space.into_serial();
        SerialPersistentHugr { state_space }
    }

    /// Create a serialized format from a reference to [`CommitStateSpace`]
    pub fn to_serial(&self) -> SerialPersistentHugr<H>
    where
        H: From<Hugr>,
    {
        let Self { state_space } = self;
        let state_space = state_space.to_serial();
        SerialPersistentHugr { state_space }
    }
}

impl<H: Into<Hugr> + From<Hugr> + serde::Serialize> From<PersistentHugr<SerdeHashResolver<H>>>
    for SerialPersistentHugr<H>
{
    fn from(value: PersistentHugr<SerdeHashResolver<H>>) -> Self {
        value.into_serial()
    }
}

impl<H: Into<Hugr> + From<Hugr> + serde::Serialize> From<SerialPersistentHugr<H>>
    for PersistentHugr<SerdeHashResolver<H>>
{
    fn from(value: SerialPersistentHugr<H>) -> Self {
        PersistentHugr::from_serial(value)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::{
        CommitId,
        tests::{WrappedHugr, test_state_space},
    };

    use rstest::rstest;

    #[rstest]
    fn test_serde_persistent_hugr(
        test_state_space: (
            CommitStateSpace<SerdeHashResolver<WrappedHugr>>,
            [CommitId; 4],
        ),
    ) {
        let (state_space, [cm1, cm2, _, cm4]) = test_state_space;

        let per_hugr = state_space.try_extract_hugr([cm1, cm2, cm4]).unwrap();
        let ser_per_hugr = per_hugr.to_serial();

        insta::assert_snapshot!(serde_json::to_string_pretty(&ser_per_hugr).unwrap());
    }
}
