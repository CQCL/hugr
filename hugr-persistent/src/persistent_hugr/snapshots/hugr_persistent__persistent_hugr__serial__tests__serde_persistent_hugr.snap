---
source: hugr-persistent/src/persistent_hugr/serial.rs
expression: "serde_json::to_string_pretty(&ser_per_hugr).unwrap()"
---
{
  "state_space": {
    "graph": {
      "nodes": {
        "4988c9cc961cdd04": {
          "value": {
            "Replacement": {
              "subgraph": {
                "nodes": [
                  [
                    "a636a4789ca929c1",
                    9
                  ],
                  [
                    "53404fc89a5b596a",
                    8
                  ]
                ],
                "inputs": [
                  [
                    [
                      [
                        "53404fc89a5b596a",
                        8
                      ],
                      {
                        "index": 0
                      }
                    ]
                  ],
                  [
                    [
                      [
                        "a636a4789ca929c1",
                        9
                      ],
                      {
                        "index": 1
                      }
                    ]
                  ]
                ],
                "outputs": [
                  [
                    [
                      "a636a4789ca929c1",
                      9
                    ],
                    {
                      "index": 0
                    }
                  ]
                ]
              },
              "replacement": {
                "hugr": "HUGRiHJv?@{\"modules\":[{\"version\":\"live\",\"nodes\":[{\"parent\":0,\"op\":\"Module\"},{\"parent\":0,\"op\":\"FuncDefn\",\"name\":\"main\",\"signature\":{\"params\":[],\"body\":{\"input\":[{\"t\":\"Sum\",\"s\":\"Unit\",\"size\":2},{\"t\":\"Sum\",\"s\":\"Unit\",\"size\":2}],\"output\":[{\"t\":\"Sum\",\"s\":\"Unit\",\"size\":2}]}},\"visibility\":\"Public\"},{\"parent\":1,\"op\":\"Input\",\"types\":[{\"t\":\"Sum\",\"s\":\"Unit\",\"size\":2},{\"t\":\"Sum\",\"s\":\"Unit\",\"size\":2}]},{\"parent\":1,\"op\":\"Output\",\"types\":[{\"t\":\"Sum\",\"s\":\"Unit\",\"size\":2}]},{\"parent\":1,\"op\":\"DFG\",\"signature\":{\"input\":[{\"t\":\"Sum\",\"s\":\"Unit\",\"size\":2},{\"t\":\"Sum\",\"s\":\"Unit\",\"size\":2}],\"output\":[{\"t\":\"Sum\",\"s\":\"Unit\",\"size\":2}]}},{\"parent\":4,\"op\":\"Input\",\"types\":[{\"t\":\"Sum\",\"s\":\"Unit\",\"size\":2},{\"t\":\"Sum\",\"s\":\"Unit\",\"size\":2}]},{\"parent\":4,\"op\":\"Output\",\"types\":[{\"t\":\"Sum\",\"s\":\"Unit\",\"size\":2}]},{\"parent\":4,\"op\":\"Extension\",\"extension\":\"logic\",\"name\":\"Xor\",\"args\":[],\"signature\":{\"input\":[{\"t\":\"Sum\",\"s\":\"Unit\",\"size\":2},{\"t\":\"Sum\",\"s\":\"Unit\",\"size\":2}],\"output\":[{\"t\":\"Sum\",\"s\":\"Unit\",\"size\":2}]}}],\"edges\":[[[2,0],[4,0]],[[2,1],[4,1]],[[4,0],[3,0]],[[5,0],[7,0]],[[5,1],[7,1]],[[7,0],[6,0]]],\"metadata\":[null,null,null,null,null,null,null,null],\"encoder\":\"hugr-rs v0.20.2\",\"entrypoint\":4}],\"extensions\":[]}"
              }
            }
          },
          "incoming": [
            [
              "a636a4789ca929c1",
              null
            ],
            [
              "53404fc89a5b596a",
              null
            ]
          ]
        },
        "53404fc89a5b596a": {
          "value": {
            "Replacement": {
              "subgraph": {
                "nodes": [
                  [
                    "a636a4789ca929c1",
                    7
                  ]
                ],
                "inputs": [
                  [
                    [
                      [
                        "a636a4789ca929c1",
                        7
                      ],
                      {
                        "index": 0
                      }
                    ]
                  ]
                ],
                "outputs": [
                  [
                    [
                      "a636a4789ca929c1",
                      7
                    ],
                    {
                      "index": 0
                    }
                  ]
                ]
              },
              "replacement": {
                "hugr": "HUGRiHJv?@{\"modules\":[{\"version\":\"live\",\"nodes\":[{\"parent\":0,\"op\":\"Module\"},{\"parent\":0,\"op\":\"FuncDefn\",\"name\":\"main\",\"signature\":{\"params\":[],\"body\":{\"input\":[{\"t\":\"Sum\",\"s\":\"Unit\",\"size\":2}],\"output\":[{\"t\":\"Sum\",\"s\":\"Unit\",\"size\":2}]}},\"visibility\":\"Public\"},{\"parent\":1,\"op\":\"Input\",\"types\":[{\"t\":\"Sum\",\"s\":\"Unit\",\"size\":2}]},{\"parent\":1,\"op\":\"Output\",\"types\":[{\"t\":\"Sum\",\"s\":\"Unit\",\"size\":2}]},{\"parent\":1,\"op\":\"DFG\",\"signature\":{\"input\":[{\"t\":\"Sum\",\"s\":\"Unit\",\"size\":2}],\"output\":[{\"t\":\"Sum\",\"s\":\"Unit\",\"size\":2}]}},{\"parent\":4,\"op\":\"Input\",\"types\":[{\"t\":\"Sum\",\"s\":\"Unit\",\"size\":2}]},{\"parent\":4,\"op\":\"Output\",\"types\":[{\"t\":\"Sum\",\"s\":\"Unit\",\"size\":2}]},{\"parent\":4,\"op\":\"Extension\",\"extension\":\"logic\",\"name\":\"Not\",\"args\":[],\"signature\":{\"input\":[{\"t\":\"Sum\",\"s\":\"Unit\",\"size\":2}],\"output\":[{\"t\":\"Sum\",\"s\":\"Unit\",\"size\":2}]}},{\"parent\":4,\"op\":\"Extension\",\"extension\":\"logic\",\"name\":\"Not\",\"args\":[],\"signature\":{\"input\":[{\"t\":\"Sum\",\"s\":\"Unit\",\"size\":2}],\"output\":[{\"t\":\"Sum\",\"s\":\"Unit\",\"size\":2}]}}],\"edges\":[[[2,0],[4,0]],[[4,0],[3,0]],[[5,0],[7,0]],[[7,0],[8,0]],[[8,0],[6,0]]],\"metadata\":[null,null,null,null,null,null,null,null,null],\"encoder\":\"hugr-rs v0.20.2\",\"entrypoint\":4}],\"extensions\":[]}"
              }
            }
          },
          "incoming": [
            [
              "a636a4789ca929c1",
              null
            ]
          ]
        },
        "9c7867f6a8cc74c8": {
          "value": {
            "Replacement": {
              "subgraph": {
                "nodes": [
                  [
                    "a636a4789ca929c1",
                    8
                  ]
                ],
                "inputs": [
                  [
                    [
                      [
                        "a636a4789ca929c1",
                        8
                      ],
                      {
                        "index": 0
                      }
                    ]
                  ]
                ],
                "outputs": [
                  [
                    [
                      "a636a4789ca929c1",
                      8
                    ],
                    {
                      "index": 0
                    }
                  ]
                ]
              },
              "replacement": {
                "hugr": "HUGRiHJv?@{\"modules\":[{\"version\":\"live\",\"nodes\":[{\"parent\":0,\"op\":\"Module\"},{\"parent\":0,\"op\":\"FuncDefn\",\"name\":\"main\",\"signature\":{\"params\":[],\"body\":{\"input\":[{\"t\":\"Sum\",\"s\":\"Unit\",\"size\":2}],\"output\":[{\"t\":\"Sum\",\"s\":\"Unit\",\"size\":2}]}},\"visibility\":\"Public\"},{\"parent\":1,\"op\":\"Input\",\"types\":[{\"t\":\"Sum\",\"s\":\"Unit\",\"size\":2}]},{\"parent\":1,\"op\":\"Output\",\"types\":[{\"t\":\"Sum\",\"s\":\"Unit\",\"size\":2}]},{\"parent\":1,\"op\":\"DFG\",\"signature\":{\"input\":[{\"t\":\"Sum\",\"s\":\"Unit\",\"size\":2}],\"output\":[{\"t\":\"Sum\",\"s\":\"Unit\",\"size\":2}]}},{\"parent\":4,\"op\":\"Input\",\"types\":[{\"t\":\"Sum\",\"s\":\"Unit\",\"size\":2}]},{\"parent\":4,\"op\":\"Output\",\"types\":[{\"t\":\"Sum\",\"s\":\"Unit\",\"size\":2}]},{\"parent\":4,\"op\":\"Extension\",\"extension\":\"logic\",\"name\":\"Not\",\"args\":[],\"signature\":{\"input\":[{\"t\":\"Sum\",\"s\":\"Unit\",\"size\":2}],\"output\":[{\"t\":\"Sum\",\"s\":\"Unit\",\"size\":2}]}},{\"parent\":4,\"op\":\"Extension\",\"extension\":\"logic\",\"name\":\"Not\",\"args\":[],\"signature\":{\"input\":[{\"t\":\"Sum\",\"s\":\"Unit\",\"size\":2}],\"output\":[{\"t\":\"Sum\",\"s\":\"Unit\",\"size\":2}]}}],\"edges\":[[[2,0],[4,0]],[[4,0],[3,0]],[[5,0],[7,0]],[[7,0],[8,0]],[[8,0],[6,0]]],\"metadata\":[null,null,null,null,null,null,null,null,null],\"encoder\":\"hugr-rs v0.20.2\",\"entrypoint\":4}],\"extensions\":[]}"
              }
            }
          },
          "incoming": [
            [
              "a636a4789ca929c1",
              null
            ]
          ]
        },
        "a636a4789ca929c1": {
          "value": {
            "Base": {
              "hugr": "HUGRiHJv?@{\"modules\":[{\"version\":\"live\",\"nodes\":[{\"parent\":0,\"op\":\"Module\"},{\"parent\":0,\"op\":\"FuncDefn\",\"name\":\"main\",\"signature\":{\"params\":[],\"body\":{\"input\":[{\"t\":\"Sum\",\"s\":\"Unit\",\"size\":2},{\"t\":\"Sum\",\"s\":\"Unit\",\"size\":2}],\"output\":[{\"t\":\"Sum\",\"s\":\"Unit\",\"size\":2}]}},\"visibility\":\"Public\"},{\"parent\":1,\"op\":\"Input\",\"types\":[{\"t\":\"Sum\",\"s\":\"Unit\",\"size\":2},{\"t\":\"Sum\",\"s\":\"Unit\",\"size\":2}]},{\"parent\":1,\"op\":\"Output\",\"types\":[{\"t\":\"Sum\",\"s\":\"Unit\",\"size\":2}]},{\"parent\":1,\"op\":\"DFG\",\"signature\":{\"input\":[{\"t\":\"Sum\",\"s\":\"Unit\",\"size\":2},{\"t\":\"Sum\",\"s\":\"Unit\",\"size\":2}],\"output\":[{\"t\":\"Sum\",\"s\":\"Unit\",\"size\":2}]}},{\"parent\":4,\"op\":\"Input\",\"types\":[{\"t\":\"Sum\",\"s\":\"Unit\",\"size\":2},{\"t\":\"Sum\",\"s\":\"Unit\",\"size\":2}]},{\"parent\":4,\"op\":\"Output\",\"types\":[{\"t\":\"Sum\",\"s\":\"Unit\",\"size\":2}]},{\"parent\":4,\"op\":\"Extension\",\"extension\":\"logic\",\"name\":\"Not\",\"args\":[],\"signature\":{\"input\":[{\"t\":\"Sum\",\"s\":\"Unit\",\"size\":2}],\"output\":[{\"t\":\"Sum\",\"s\":\"Unit\",\"size\":2}]}},{\"parent\":4,\"op\":\"Extension\",\"extension\":\"logic\",\"name\":\"Not\",\"args\":[],\"signature\":{\"input\":[{\"t\":\"Sum\",\"s\":\"Unit\",\"size\":2}],\"output\":[{\"t\":\"Sum\",\"s\":\"Unit\",\"size\":2}]}},{\"parent\":4,\"op\":\"Extension\",\"extension\":\"logic\",\"name\":\"And\",\"args\":[],\"signature\":{\"input\":[{\"t\":\"Sum\",\"s\":\"Unit\",\"size\":2},{\"t\":\"Sum\",\"s\":\"Unit\",\"size\":2}],\"output\":[{\"t\":\"Sum\",\"s\":\"Unit\",\"size\":2}]}}],\"edges\":[[[2,0],[4,0]],[[2,1],[4,1]],[[4,0],[3,0]],[[5,0],[7,0]],[[5,1],[8,0]],[[7,0],[9,0]],[[8,0],[9,1]],[[9,0],[6,0]]],\"metadata\":[null,null,null,null,null,null,null,null,null,null],\"encoder\":\"hugr-rs v0.20.2\",\"entrypoint\":4}],\"extensions\":[]}"
            }
          },
          "incoming": []
        }
      },
      "resolver_id": {
        "id": "SerdeHashResolver"
      }
    },
    "base_commit": "a636a4789ca929c1"
  }
}
